File: SI3/S6/dsa/td7/src/tools/GraphHelper.java
package graphs.tools.ORIGIN;

import graphs.graph.*;

import java.util.*;

/**
 * A class to find cycles in undirected and directed graphs
 */
public class GraphHelper {
	
	private  Graph aGraph;


	/////////////// Cycle detection for undirected graphs

	public GraphHelper() {
		//TO IMPLEMENT
	}

	/**
	 * Returns true if the graph 'G' is cyclic
	 * false otherwise
	 */
	public boolean hasCycle(UnDiGraph unDiGraph) {
		return false;
	}
	
	/**
	 * Returns true if a cycle was found by traversing
	 * the graph, coming from vertex from and going by vertex u
	 * Precondition: vertex 'from' is visited and vertex 'u' is
	 * not visited
	 */
	private boolean hasCycle(Vertex u, Vertex from) {
		return false;
	}
	
	/////////////// Cycle detection for directed graphs
	
	private enum Status {UN_DISCOVERED, IN_PROGRESS, COMPLETED} // status of vertex

	/**
	 * Returns true if the graph 'G' is cyclic
	 * false otherwise
	 */
	public boolean hasCycle(DiGraph diGraph) {
		return false;
	}
	
	/**
	 * Returns true if a cycle was found by traversing
	 * the graph from vertex u
	 * Precondition: vertex 'u' is 'UnDiscovered'
	 */
	private boolean hasCycle(Vertex u) {
		return false;
	}
	
	

	/* ------------------- Path finding ------------------- */

	/**
	 * Returns a path as from vertex 'u' to vertex 'v' in the graph 'G'
	 * as a list of vertices if such a path exists, the empty list otherwise
	 */
	public List<Vertex> findPath(Graph graph, Vertex u, Vertex v) {
		//TO IMPLEMENT
		List<Vertex> path = new LinkedList<>();
		return path;
	}

	/* ------------------- Root finding ------------------- */

	/**
	 * Returns a root of the graph 'G' if
	 * such root exists, null otherwise
	 */
	public Optional<Vertex> findRoot(DiGraph diGraph) {
		return Optional.empty();
	}


	public  double computeDistanceOfPath(List<Vertex> path, AbstractGraph graph) {
		return 0.0;
	}

	/////////////// 

}

File: SI3/S6/dsa/td7/src/tools/GraphHelper.kt
package graphs.tools

import graphs.AbstractGraph
import graphs.DiGraph
import graphs.Graph
import graphs.Vertex
import java.util.*

/**
 * A class to find cycles in undirected and directed graphs
 */
class GraphHelper  /////////////// Cycle detection for undirected graphs
{
    fun hasCycle(graph: Graph): Boolean {
        val visited = ArrayList<Vertex>()
        for (vertex in graph.vertices()) {
            if (!visited.contains(vertex) && hasCycleUtil(graph, vertex, visited, null)) {
                return true
            }
        }
        return false
    }

    private fun hasCycleUtil(graph: Graph, current: Vertex, visited: ArrayList<Vertex>, parent: Vertex?): Boolean {
        visited.add(current)
        for (adj in graph.adjacents(current)) {
            if (!visited.contains(adj)) {
                if (hasCycleUtil(graph, adj, visited, current)) {
                    return true
                }
            } else if (adj != parent) {
                return true
            }
        }
        return false
    }

    /**
     * Returns a path as from vertex 'u' to vertex 'v' in the graph 'G'
     * as a list of vertices if such a path exists, the empty list otherwise
     */
    fun findPath(graph: Graph, u: Vertex, v: Vertex): List<Vertex> {
        val path = dfs(graph, u)
        if (path.contains(v)) {
            val index = path.indexOf(v)
            return path.subList(0, index + 1)
        }
        return ArrayList()
    }
    /* ------------------- Path finding ------------------- */
    /**
     * Returns a root of the graph 'G' if
     * such root exists, null otherwise
     */
    fun findRoot(diGraph: DiGraph): Optional<Vertex> {
        return Optional.empty()
    }

    /* ------------------- Root finding ------------------- */
    fun computeDistanceOfPath(path: List<Vertex>, graph: AbstractGraph): Double {
        var distance = 0.0
        for (i in 0 until path.size - 1) {
            val u = path[i]
            val v = path[i + 1]
            if (!graph.adjacents(u).contains(v)) {
                return -1.0
            }
            distance += graph.findEdge(u, v)!!.weight()
        }
        return distance
    }

    private enum class Status {
        UN_DISCOVERED,
        IN_PROGRESS,
        COMPLETED
    } // status of vertex
    ///////////////
}
